File: neomatrix-core/src/math/losses.rs
//! Loss (cost) functions for neural network training.
//!
//! This module provides the `LossFunction` trait and implementations for common loss functions
//! used in supervised learning. Each loss function measures the discrepancy between predicted
//! outputs and true labels, with their derivatives used for backpropagation during training.
//!
//! # Available Loss Functions
//!
//! - **MSE** (`MeanSquaredError`): L2 loss - penalizes large errors quadratically
//! - **MAE** (`MeanAbsoluteError`): L1 loss - robust to outliers, linear penalty
//! - **BCE** (`BinaryCrossEntropy`): For binary classification with sigmoid output
//! - **CCE** (`CategoricalCrossEntropy`): For multi-class classification with softmax output
//! - **Huber Loss** (`HuberLoss`): Hybrid MSE/MAE - quadratic for small errors, linear for large
//! - **Hinge Loss** (`HingeLoss`): For SVM-style margin-based classification
//!
//! # Performance
//!
//! All loss functions leverage Rayon's parallel iterators for batch processing, automatically
//! distributing computation across available CPU cores when processing multiple samples.
//!
//! # Numerical Stability
//!
//! Cross-entropy losses (BCE, CCE) use epsilon clamping (`1e-7`) to prevent `log(0)` and
//! division by zero errors during gradient computation.

use crate::errors::MathError;
use crate::tensor::Tensor;
use ndarray::parallel::prelude::*;
use ndarray::{Axis, Zip};

/// Trait defining the interface for loss functions.
///
/// All loss functions must implement this trait to be used for training neural networks.
/// Both methods operate on tensors representing batches of samples.
///
/// # Parameters
///
/// - `t`: True labels (ground truth)
/// - `z`: Predicted outputs from the model
///
/// # Methods
///
/// - `function`: Computes the scalar loss value for a batch
/// - `derivative`: Computes the gradient w.r.t. predictions for backpropagation
///
/// # Thread Safety
///
/// The `Send + Sync` bounds allow loss functions to be safely shared across threads,
/// enabling parallel batch processing during training.
///
/// # Shape Requirements
///
/// Both `t` and `z` must have identical shapes, or an error is returned. The first
/// dimension is typically the batch dimension.
pub trait LossFunction: Send + Sync {
    /// Computes the loss value for a batch of samples.
    ///
    /// # Parameters
    ///
    /// - `t`: True labels tensor (shape: `[batch_size, ...]`)
    /// - `z`: Predicted outputs tensor (shape: `[batch_size, ...]`)
    ///
    /// # Returns
    ///
    /// A scalar f32 representing the average loss across the batch, or an error
    /// if shapes don't match or computation fails.
    fn function(&self, t: &Tensor, z: &Tensor) -> Result<f32, MathError>;

    /// Computes the gradient of the loss w.r.t. predictions.
    ///
    /// # Parameters
    ///
    /// - `t`: True labels tensor
    /// - `z`: Predicted outputs tensor
    ///
    /// # Returns
    ///
    /// A tensor with the same shape as the input, containing the gradient values
    /// for backpropagation, or an error if shapes don't match.
    fn derivative(&self, t: &Tensor, z: &Tensor) -> Result<Tensor, MathError>;
}

// ─────────────────────────────────────────────────────────────────────────────
// Implementations
// ─────────────────────────────────────────────────────────────────────────────

/// Mean Squared Error (MSE) loss function.
///
/// # Mathematical Definition
///
/// - **Function**: `L(t,z) = (1/n) * Σ(t_i - z_i)²`
/// - **Derivative**: `∂L/∂z_i = -2(t_i - z_i) / n`
///
/// # Properties
///
/// - L2 loss - penalizes large errors quadratically
/// - Sensitive to outliers due to squaring
/// - Commonly used for regression tasks
/// - Differentiable everywhere
///
/// # Use Cases
///
/// - Regression problems
/// - When large errors should be heavily penalized
/// - Gaussian noise assumptions
pub struct MeanSquaredError;

impl LossFunction for MeanSquaredError {
    fn function(&self, t: &Tensor, z: &Tensor) -> Result<f32, MathError> {
        if t.shape != z.shape {
            return Err(MathError::CostFunctionBatchShapeMismatch);
        }

        // Parallel computation across batch dimension
        let sum: f32 = t
            .data
            .axis_iter(Axis(0))
            .into_par_iter()
            .zip(z.data.axis_iter(Axis(0)).into_par_iter())
            .map(|(t_i, z_i)| -> Result<f32, MathError> {
                let t_i = Tensor {
                    dimension: t_i.ndim(),
                    shape: t_i.shape().to_vec(),
                    data: t_i.to_owned(),
                };
                let z_i = Tensor {
                    dimension: z_i.ndim(),
                    shape: z_i.shape().to_vec(),
                    data: z_i.to_owned(),
                };
                // Σ(t_i - z_i)²
                let diff = (&t_i - &z_i).map_err(|_| MathError::TensorSubtractionFailed)?;
                Ok(diff.data.mapv(|x| x.powi(2)).sum())
            })
            .collect::<Result<Vec<f32>, MathError>>()?
            .into_iter()
            .sum();

        // (1/n) * Σ(...)
        Ok(sum / (t.length() as f32))
    }

    fn derivative(&self, t: &Tensor, z: &Tensor) -> Result<Tensor, MathError> {
        if t.shape != z.shape {
            return Err(MathError::DerivativeShapeMismatch);
        }
        let n = t.length() as f32;
        let diff = (t - z).map_err(|_| MathError::TensorSubtractionFailed)?;
        // ∂L/∂z_i = -2(t_i - z_i) / n
        let gradients = diff.data.mapv(|x| -x * 2.0 / n);
        Ok(Tensor {
            dimension: gradients.ndim(),
            shape: gradients.shape().to_vec(),
            data: gradients,
        })
    }
}

/// Mean Absolute Error (MAE) loss function.
///
/// # Mathematical Definition
///
/// - **Function**: `L(t,z) = (1/n) * Σ|t_i - z_i|`
/// - **Derivative**: `∂L/∂z_i = -sign(t_i - z_i) / n`
///
/// # Properties
///
/// - L1 loss - penalizes errors linearly
/// - More robust to outliers than MSE
/// - Not differentiable at zero (uses sign function)
/// - Gradients have constant magnitude regardless of error size
///
/// # Use Cases
///
/// - Regression with outliers in data
/// - When all errors should be weighted equally
/// - Robust optimization scenarios
pub struct MeanAbsoluteError;

impl LossFunction for MeanAbsoluteError {
    fn function(&self, t: &Tensor, z: &Tensor) -> Result<f32, MathError> {
        if t.shape != z.shape {
            return Err(MathError::CostFunctionBatchShapeMismatch);
        }

        // Parallel computation across batch dimension
        let sum: f32 = t
            .data
            .axis_iter(Axis(0))
            .into_par_iter()
            .zip(z.data.axis_iter(Axis(0)).into_par_iter())
            .map(|(t_i, z_i)| -> Result<f32, MathError> {
                let t_i = Tensor {
                    dimension: t_i.ndim(),
                    shape: t_i.shape().to_vec(),
                    data: t_i.to_owned(),
                };
                let z_i = Tensor {
                    dimension: z_i.ndim(),
                    shape: z_i.shape().to_vec(),
                    data: z_i.to_owned(),
                };
                // Σ|t_i - z_i|
                let diff = (&t_i - &z_i).map_err(|_| MathError::TensorSubtractionFailed)?;
                Ok(diff.data.mapv(|x| x.abs()).sum())
            })
            .collect::<Result<Vec<f32>, MathError>>()?
            .into_iter()
            .sum();

        // (1/n) * Σ|...|
        Ok(sum / (t.length() as f32))
    }

    fn derivative(&self, t: &Tensor, z: &Tensor) -> Result<Tensor, MathError> {
        if t.shape != z.shape {
            return Err(MathError::DerivativeShapeMismatch);
        }
        let diff = (t - z).map_err(|_| MathError::TensorSubtractionFailed)?;
        // ∂L/∂z_i = -sign(t_i - z_i) / n
        // At zero: gradient is undefined, we return 0
        let gradients = diff.data.mapv(|x| {
            if x == 0.0 {
                0.0
            } else {
                -(x.abs() / x) / t.length() as f32
            }
        });
        Ok(Tensor {
            dimension: gradients.ndim(),
            shape: gradients.shape().to_vec(),
            data: gradients,
        })
    }
}

/// Binary Cross-Entropy (BCE) loss function.
///
/// # Mathematical Definition
///
/// - **Function**: `L(t,z) = -(1/n) * Σ[t_i·log(z_i) + (1-t_i)·log(1-z_i)]`
/// - **Derivative**: `∂L/∂z_i = -[t_i/z_i - (1-t_i)/(1-z_i)] / n`
///
/// # Properties
///
/// - Designed for binary classification tasks
/// - Expects labels in {0, 1} and predictions in (0, 1)
/// - Typically used with sigmoid activation in output layer
/// - Numerically stabilized with epsilon clamping to prevent `log(0)`
///
/// # Numerical Stability
///
/// Predictions are clamped to `[1e-7, 1-1e-7]` to avoid `log(0)` and division by zero.
///
/// # Use Cases
///
/// - Binary classification problems
/// - Multi-label classification (independent binary decisions per label)
/// - When output layer uses sigmoid activation
pub struct BinaryCrossEntropy;

impl LossFunction for BinaryCrossEntropy {
    fn function(&self, t: &Tensor, z: &Tensor) -> Result<f32, MathError> {
        if t.shape != z.shape {
            return Err(MathError::CostFunctionBatchShapeMismatch);
        }
        // Clamp predictions to prevent log(0)
        let epsilon = 1e-7;

        // Parallel computation across batch dimension
        let sum: f32 = t
            .data
            .axis_iter(Axis(0))
            .into_par_iter()
            .zip(z.data.axis_iter(Axis(0)).into_par_iter())
            .map(|(t_i, z_i)| -> Result<f32, MathError> {
                // L = -Σ[t·log(z) + (1-t)·log(1-z)]
                let sum: f32 = t_i
                    .iter()
                    .zip(z_i.iter())
                    .map(|(t_j, z_j)| {
                        let z_clamped = z_j.clamp(epsilon, 1.0 - epsilon);
                        t_j * z_clamped.ln() + (1.0 - t_j) * (1.0 - z_clamped).ln()
                    })
                    .sum();
                Ok(-sum)
            })
            .collect::<Result<Vec<f32>, MathError>>()?
            .into_iter()
            .sum();

        // (1/n) * Σ(...)
        Ok(sum / (t.length() as f32))
    }

    fn derivative(&self, t: &Tensor, z: &Tensor) -> Result<Tensor, MathError> {
        if t.shape != z.shape {
            return Err(MathError::DerivativeShapeMismatch);
        }

        let epsilon = 1e-7;

        // ∂L/∂z = -[t/z - (1-t)/(1-z)] / n
        let gradients = Zip::from(&t.data).and(&z.data).par_map_collect(|t_i, z_i| {
            let z_clamped = z_i.clamp(epsilon, 1.0 - epsilon);
            -((t_i / z_clamped) - ((1.0 - t_i) / (1.0 - z_clamped))) / t.length() as f32
        });

        Ok(Tensor {
            dimension: gradients.ndim(),
            shape: gradients.shape().to_vec(),
            data: gradients,
        })
    }
}

/// Categorical Cross-Entropy (CCE) loss function.
///
/// # Mathematical Definition
///
/// - **Function**: `L(t,z) = -(1/n) * ΣΣ t_i,k·log(z_i,k)`
/// - **Derivative**: `∂L/∂z_i,k = -t_i,k / z_i,k / n`
///
/// # Properties
///
/// - Designed for multi-class classification (mutually exclusive classes)
/// - Expects one-hot encoded labels: `t_i,k ∈ {0,1}` with `Σt_i,k = 1`
/// - Expects softmax-normalized predictions: `z_i,k ∈ (0,1)` with `Σz_i,k = 1`
/// - Typically used with softmax activation in output layer
/// - Numerically stabilized with epsilon clamping
///
/// # Numerical Stability
///
/// Predictions are clamped to `[1e-7, 1-1e-7]` to avoid `log(0)`.
///
/// # Use Cases
///
/// - Multi-class classification (single label per sample)
/// - When output layer uses softmax activation
/// - Image classification, NLP classification tasks
pub struct CategoricalCrossEntropy;

impl LossFunction for CategoricalCrossEntropy {
    fn function(&self, t: &Tensor, z: &Tensor) -> Result<f32, MathError> {
        if t.shape != z.shape {
            return Err(MathError::CostFunctionBatchShapeMismatch);
        }
        // Clamp predictions to prevent log(0)
        let epsilon = 1e-7;

        // Parallel computation across batch dimension
        let sum: f32 = t
            .data
            .axis_iter(Axis(0))
            .into_par_iter()
            .zip(z.data.axis_iter(Axis(0)).into_par_iter())
            .map(|(t_i, z_i)| -> Result<f32, MathError> {
                // L = -Σ t_k·log(z_k)
                let sum: f32 = t_i
                    .iter()
                    .zip(z_i.iter())
                    .map(|(t_j, z_j)| {
                        let z_clamped = z_j.clamp(epsilon, 1.0 - epsilon);
                        t_j * z_clamped.ln()
                    })
                    .sum();
                Ok(-sum)
            })
            .collect::<Result<Vec<f32>, MathError>>()?
            .into_iter()
            .sum();

        // (1/n) * Σ(...)
        Ok(sum / (t.length() as f32))
    }

    fn derivative(&self, t: &Tensor, z: &Tensor) -> Result<Tensor, MathError> {
        if t.shape != z.shape {
            return Err(MathError::DerivativeShapeMismatch);
        }

        let epsilon = 1e-7;

        // ∂L/∂z = -t / z / n
        let gradients = Zip::from(&t.data).and(&z.data).par_map_collect(|t_i, z_i| {
            let z_clamped = z_i.clamp(epsilon, 1.0 - epsilon);
            -(t_i / z_clamped) / t.length() as f32
        });

        Ok(Tensor {
            dimension: gradients.ndim(),
            shape: gradients.shape().to_vec(),
            data: gradients,
        })
    }
}

/// Huber Loss function.
///
/// # Mathematical Definition
///
/// ```text
/// L(t,z) = (1/n) * Σ L_δ(t_i - z_i)
///
/// where L_δ(x) = {
///   0.5·x²          if |x| ≤ δ
///   δ·|x| - 0.5·δ²  if |x| > δ
/// }
/// ```
///
/// Derivative:
/// ```text
/// ∂L/∂z_i = {
///   (z_i - t_i) / n           if |t_i - z_i| ≤ δ
///   -δ·sign(t_i - z_i) / n    if |t_i - z_i| > δ
/// }
/// ```
///
/// # Properties
///
/// - Hybrid loss combining MSE and MAE characteristics
/// - Quadratic for small errors (|x| ≤ δ): smooth, differentiable
/// - Linear for large errors (|x| > δ): robust to outliers
/// - Transition point controlled by delta parameter (δ)
/// - More robust than MSE, smoother gradient than MAE
///
/// # Parameters
///
/// - `delta`: Threshold determining quadratic/linear transition point
///
/// # Use Cases
///
/// - Regression with occasional outliers
/// - When you want MSE smoothness for typical errors but MAE robustness for outliers
/// - Object detection bounding box regression
pub struct HuberLoss {
    /// Threshold δ for switching between quadratic and linear loss.
    pub delta: f32,
}

impl LossFunction for HuberLoss {
    fn function(&self, t: &Tensor, z: &Tensor) -> Result<f32, MathError> {
        if t.shape != z.shape {
            return Err(MathError::CostFunctionBatchShapeMismatch);
        }

        // Parallel computation across batch dimension
        let sum: f32 = t
            .data
            .axis_iter(Axis(0))
            .into_par_iter()
            .zip(z.data.axis_iter(Axis(0)).into_par_iter())
            .map(|(t_i, z_i)| -> Result<f32, MathError> {
                let sum: f32 = t_i
                    .iter()
                    .zip(z_i.iter())
                    .map(|(t_j, z_j)| {
                        let diff = t_j - z_j;
                        // L_δ(x) = 0.5·x² if |x| ≤ δ, else δ·|x| - 0.5·δ²
                        if diff.abs() <= self.delta {
                            diff.powi(2) / 2.0
                        } else {
                            self.delta * diff.abs() - self.delta.powi(2) / 2.0
                        }
                    })
                    .sum();
                Ok(sum)
            })
            .collect::<Result<Vec<f32>, MathError>>()?
            .into_iter()
            .sum();

        // (1/n) * Σ(...)
        Ok(sum / (t.length() as f32))
    }

    fn derivative(&self, t: &Tensor, z: &Tensor) -> Result<Tensor, MathError> {
        if t.shape != z.shape {
            return Err(MathError::DerivativeShapeMismatch);
        }

        // ∂L/∂z = (z - t)/n if |t - z| ≤ δ, else -δ·sign(t - z)/n
        let gradients = Zip::from(&t.data).and(&z.data).par_map_collect(|t_i, z_i| {
            let diff = t_i - z_i;
            if diff.abs() <= self.delta {
                (z_i - t_i) / t.length() as f32
            } else {
                (-self.delta * diff.abs() / diff) / t.length() as f32
            }
        });

        Ok(Tensor {
            dimension: gradients.ndim(),
            shape: gradients.shape().to_vec(),
            data: gradients,
        })
    }
}

/// Hinge Loss function.
///
/// # Mathematical Definition
///
/// - **Function**: `L(t,z) = (1/n) * Σ max(0, 1 - t_i·z_i)`
/// - **Derivative**: `∂L/∂z_i = -t_i / n` if `t_i·z_i < 1`, else `0`
///
/// # Properties
///
/// - Designed for binary classification with margin-based learning
/// - Expects labels in {-1, +1} (not {0, 1})
/// - Penalizes predictions that are on the wrong side or too close to the decision boundary
/// - Zero loss when correct prediction with sufficient margin (`t·z ≥ 1`)
/// - Used in Support Vector Machines (SVMs)
///
/// # Margin Interpretation
///
/// - `t·z > 1`: Correct with margin → loss = 0
/// - `0 < t·z < 1`: Correct but insufficient margin → linear penalty
/// - `t·z < 0`: Incorrect prediction → larger linear penalty
///
/// # Use Cases
///
/// - Binary classification with margin enforcement
/// - Support Vector Machines (SVMs)
/// - When you want to encourage confident predictions
pub struct HingeLoss;

impl LossFunction for HingeLoss {
    fn function(&self, t: &Tensor, z: &Tensor) -> Result<f32, MathError> {
        if t.shape != z.shape {
            return Err(MathError::CostFunctionBatchShapeMismatch);
        }

        // Parallel computation across batch dimension
        let sum: f32 = t
            .data
            .axis_iter(Axis(0))
            .into_par_iter()
            .zip(z.data.axis_iter(Axis(0)).into_par_iter())
            .map(|(t_i, z_i)| -> Result<f32, MathError> {
                // L = Σ max(0, 1 - t·z)
                let sum: f32 = t_i
                    .iter()
                    .zip(z_i.iter())
                    .map(|(t_j, z_j)| 0.0_f32.max(1.0 - t_j * z_j))
                    .sum();
                Ok(sum)
            })
            .collect::<Result<Vec<f32>, MathError>>()?
            .into_iter()
            .sum();

        // (1/n) * Σ(...)
        Ok(sum / (t.length() as f32))
    }

    fn derivative(&self, t: &Tensor, z: &Tensor) -> Result<Tensor, MathError> {
        if t.shape != z.shape {
            return Err(MathError::DerivativeShapeMismatch);
        }

        // ∂L/∂z = -t/n if t·z < 1, else 0
        let gradients = Zip::from(&t.data).and(&z.data).par_map_collect(|t_i, z_i| {
            let x = if t_i * z_i < 1.0 { -t_i } else { 0.0 };
            x / t.length() as f32
        });

        Ok(Tensor {
            dimension: gradients.ndim(),
            shape: gradients.shape().to_vec(),
            data: gradients,
        })
    }
}

File: neomatrix/losses.py
"""
Loss functions exposed by the NeoMatrix Rust backend.

Available: MSE, MAE, BCE, CCE, HuberLoss, HingeLoss.
"""

from typing import TYPE_CHECKING, Protocol, runtime_checkable

if TYPE_CHECKING:
    from neomatrix._backend import Tensor

from neomatrix._backend import (
    MSE,
    MAE,
    BCE,
    CCE,
    HuberLoss,
    HingeLoss,
)

__all__ = [
    "MSE",
    "MAE",
    "BCE",
    "CCE",
    "HuberLoss",
    "HingeLoss",
    # Protocols
    "LossFunction",
    "FusedLossFunction",
]


@runtime_checkable
class LossFunction(Protocol):
    """Structural protocol for any loss function.

    All loss functions expose ``call`` (forward pass, returns scalar)
    and ``backward`` (gradient w.r.t. predictions).
    """

    def call(self, y_true: Tensor, y_pred: Tensor) -> float: ...
    def backward(self, y_true: Tensor, y_pred: Tensor) -> Tensor: ...


@runtime_checkable
class FusedLossFunction(LossFunction, Protocol):
    """Structural protocol for loss functions that support fused gradients.

    ``BCE`` and ``CCE`` expose ``backward_optimized``, which returns the
    simplified gradient ``y_pred - y_true`` (valid when the preceding
    activation is Sigmoid/Softmax respectively).
    """

    def backward_optimized(self, y_true: Tensor, y_pred: Tensor) -> Tensor: ...


# ============================================================================
# LOSS FUNCTION DOCSTRINGS
# ============================================================================

MSE.__doc__ = """
Mean Squared Error loss function.

**Formula:**
    L = (1/n) Σᵢ (yᵢ - ŷᵢ)²

**Gradient (w.r.t. predictions):**
    ∂L/∂ŷ = 2(ŷ - y) / n

**Constructor:**
    ```python
    MSE()
    ```

**Parameters:**
    None (no configuration required).

**Methods:**
    - `call(y_true: Tensor, y_pred: Tensor) -> float`: Compute loss value (scalar)
    - `backward(y_true: Tensor, y_pred: Tensor) -> Tensor`: Compute gradient w.r.t. predictions

**Example:**
    ```python
    from neomatrix import losses
    from neomatrix._backend import Tensor
    
    loss_fn = losses.MSE()
    
    # Forward pass
    y_true = Tensor.from_numpy(np.array([[1.0, 2.0, 3.0]]))
    y_pred = Tensor.from_numpy(np.array([[1.1, 1.9, 3.2]]))
    loss_val = loss_fn.call(y_true, y_pred)  # Returns scalar
    
    # Backward pass
    grad = loss_fn.backward(y_true, y_pred)  # Shape: [1, 3]
    ```

**Use Cases:**
    - Regression tasks (continuous output)
    - Metric: penalizes large errors quadratically
    - Sensitive to outliers (use MAE or Huber for robustness)

**Properties:**
    - Differentiable everywhere
    - Convex optimization landscape
    - Range: [0, ∞)
"""

MAE.__doc__ = """
Mean Absolute Error loss function.

**Formula:**
    L = (1/n) Σᵢ |yᵢ - ŷᵢ|

**Gradient (w.r.t. predictions):**
    ∂L/∂ŷ = -sign(y - ŷ) / n

**Constructor:**
    ```python
    MAE()
    ```

**Parameters:**
    None (no configuration required).

**Methods:**
    - `call(y_true: Tensor, y_pred: Tensor) -> float`: Compute loss value (scalar)
    - `backward(y_true: Tensor, y_pred: Tensor) -> Tensor`: Compute gradient w.r.t. predictions

**Example:**
    ```python
    from neomatrix import losses
    from neomatrix._backend import Tensor
    
    loss_fn = losses.MAE()
    
    # Forward pass
    y_true = Tensor.from_numpy(np.array([[1.0, 2.0, 3.0]]))
    y_pred = Tensor.from_numpy(np.array([[1.5, 1.8, 3.5]]))
    loss_val = loss_fn.call(y_true, y_pred)  # Returns scalar
    
    # Backward pass
    grad = loss_fn.backward(y_true, y_pred)  # Shape: [1, 3]
    ```

**Use Cases:**
    - Robust regression (less sensitive to outliers than MSE)
    - Metric: penalizes errors linearly
    - Preferred when large errors should not dominate training

**Properties:**
    - Differentiable everywhere except at zero (gradient uses sign function)
    - Convex optimization landscape
    - Range: [0, ∞)
    - More robust to outliers than MSE
"""

BCE.__doc__ = """
Binary Cross-Entropy loss function for binary classification.

**Formula:**
    L = -(1/n) Σᵢ [yᵢ·log(ŷᵢ) + (1-yᵢ)·log(1-ŷᵢ)]

**Gradient (w.r.t. predictions):**
    ∂L/∂ŷ = (ŷ - y) / (ŷ·(1-ŷ)·n)

**Optimized Gradient (fused with Sigmoid activation):**
    When used with Sigmoid output layer:
    ∂L/∂z = σ(z) - y  (where z = pre-activation logits)

**Constructor:**
    ```python
    BCE()
    ```

**Parameters:**
    None (no configuration required).

**Methods:**
    - `call(y_true: Tensor, y_pred: Tensor) -> float`: Compute loss value (scalar)
    - `backward(y_true: Tensor, y_pred: Tensor) -> Tensor`: Compute gradient w.r.t. predictions
    - `backward_optimized(y_true: Tensor, y_pred: Tensor) -> Tensor`: Compute optimized gradient (for Sigmoid + BCE)

**Example:**
    ```python
    from neomatrix import losses
    from neomatrix._backend import Tensor
    
    loss_fn = losses.BCE()
    
    # Standard usage (predictions already passed through Sigmoid)
    y_true = Tensor.from_numpy(np.array([[0.0, 1.0, 1.0]]))  # Binary labels
    y_pred = Tensor.from_numpy(np.array([[0.1, 0.9, 0.8]]))  # Sigmoid outputs
    loss_val = loss_fn.call(y_true, y_pred)
    grad = loss_fn.backward(y_true, y_pred)
    
    # Optimized usage (fused Sigmoid + BCE gradient)
    if isinstance(loss_fn, losses.FusedLossFunction):
        # y_pred here should be PRE-ACTIVATION logits (before Sigmoid)
        optimized_grad = loss_fn.backward_optimized(y_true, y_pred)
    ```

**Use Cases:**
    - Binary classification (two classes: 0 or 1)
    - Multi-label classification (independent binary decisions per label)
    - Output layer: Sigmoid activation

**Properties:**
    - Expects predictions in range (0, 1) — use Sigmoid activation
    - Expects labels to be 0 or 1
    - Numerically stable implementation (log(0) handled internally)
    - Fused gradient optimization available with Sigmoid

**Limitations:**
    - Sensitive to class imbalance (consider weighted BCE or focal loss)
    - Predictions exactly 0 or 1 cause numerical issues (use label smoothing)
"""

CCE.__doc__ = """
Categorical Cross-Entropy loss function for multi-class classification.

**Formula:**
    L = -(1/n) Σᵢ Σₖ yᵢₖ·log(ŷᵢₖ)

**Gradient (w.r.t. predictions):**
    ∂L/∂ŷ = -y / (ŷ·n)

**Optimized Gradient (fused with Softmax activation):**
    When used with Softmax output layer:
    ∂L/∂z = softmax(z) - y  (where z = pre-activation logits)

**Constructor:**
    ```python
    CCE()
    ```

**Parameters:**
    None (no configuration required).

**Methods:**
    - `call(y_true: Tensor, y_pred: Tensor) -> float`: Compute loss value (scalar)
    - `backward(y_true: Tensor, y_pred: Tensor) -> Tensor`: Compute gradient w.r.t. predictions
    - `backward_optimized(y_true: Tensor, y_pred: Tensor) -> Tensor`: Compute optimized gradient (for Softmax + CCE)

**Example:**
    ```python
    from neomatrix import losses
    from neomatrix._backend import Tensor
    import numpy as np
    
    loss_fn = losses.CCE()
    
    # Standard usage (predictions already passed through Softmax)
    y_true = Tensor.from_numpy(np.array([[1.0, 0.0, 0.0]]))  # One-hot encoded
    y_pred = Tensor.from_numpy(np.array([[0.8, 0.15, 0.05]]))  # Softmax outputs
    loss_val = loss_fn.call(y_true, y_pred)
    grad = loss_fn.backward(y_true, y_pred)
    
    # Optimized usage (fused Softmax + CCE gradient)
    if isinstance(loss_fn, losses.FusedLossFunction):
        # y_pred here should be PRE-ACTIVATION logits (before Softmax)
        optimized_grad = loss_fn.backward_optimized(y_true, y_pred)
    ```

**Use Cases:**
    - Multi-class classification (K mutually exclusive classes)
    - Output layer: Softmax activation
    - Single-label classification (each sample belongs to exactly one class)

**Properties:**
    - Expects predictions to sum to 1 per sample (use Softmax)
    - Expects labels to be one-hot encoded (or probability distributions)
    - Numerically stable implementation (log(0) handled internally)
    - Fused gradient optimization available with Softmax

**Limitations:**
    - Not suitable for multi-label classification (use BCE instead)
    - Sensitive to class imbalance (consider weighted CCE or focal loss)
    - Predictions exactly 0 cause numerical issues (use label smoothing)
"""

HuberLoss.__doc__ = """
Huber loss function for robust regression.

**Formula:**
    For |error| ≤ δ (quadratic region):
        L = (1/n) Σᵢ (1/2)·(yᵢ - ŷᵢ)²
    
    For |error| > δ (linear region):
        L = (1/n) Σᵢ δ·(|yᵢ - ŷᵢ| - δ/2)

**Gradient (w.r.t. predictions):**
    ∂L/∂ŷ = {
        (ŷ - y) / n,              if |y - ŷ| ≤ δ
        δ·sign(ŷ - y) / n,        if |y - ŷ| > δ
    }

**Constructor:**
    ```python
    HuberLoss(delta: float = 1.0)
    ```

**Parameters:**
    - `delta` (float): Threshold separating quadratic and linear regions. Default: 1.0.
        - Small δ → more robust to outliers (behaves like MAE)
        - Large δ → less robust, behaves like MSE

**Methods:**
    - `call(y_true: Tensor, y_pred: Tensor) -> float`: Compute loss value (scalar)
    - `backward(y_true: Tensor, y_pred: Tensor) -> Tensor`: Compute gradient w.r.t. predictions

**Example:**
    ```python
    from neomatrix import losses
    from neomatrix._backend import Tensor
    import numpy as np
    
    # Default delta = 1.0
    loss_fn = losses.HuberLoss()
    
    # Custom delta for higher robustness
    robust_loss = losses.HuberLoss(delta=0.5)
    
    # Forward pass
    y_true = Tensor.from_numpy(np.array([[1.0, 2.0, 10.0]]))  # Note: outlier at 10.0
    y_pred = Tensor.from_numpy(np.array([[1.1, 1.9, 3.0]]))
    loss_val = loss_fn.call(y_true, y_pred)
    
    # Backward pass
    grad = loss_fn.backward(y_true, y_pred)
    ```

**Use Cases:**
    - Regression with outliers (robust alternative to MSE)
    - Time series forecasting with noisy data
    - Balancing sensitivity to small vs large errors

**Properties:**
    - Differentiable everywhere
    - Convex optimization landscape
    - Quadratic near zero (like MSE) → faster convergence for small errors
    - Linear for large errors (like MAE) → robust to outliers
    - Range: [0, ∞)

**Tuning Delta:**
    - δ = 0.1: Very robust, almost pure MAE
    - δ = 1.0: Standard Huber loss (default)
    - δ = 10.0: Less robust, closer to MSE
"""

HingeLoss.__doc__ = """
Hinge loss function for SVM-style binary classification.

**Formula:**
    L = (1/n) Σᵢ max(0, 1 - yᵢ·ŷᵢ)

**Gradient (w.r.t. predictions):**
    ∂L/∂ŷ = {
        -y / n,    if y·ŷ < 1 (margin violation)
        0,         if y·ŷ ≥ 1 (correct classification with margin)
    }

**Constructor:**
    ```python
    HingeLoss()
    ```

**Parameters:**
    None (no configuration required).

**Methods:**
    - `call(y_true: Tensor, y_pred: Tensor) -> float`: Compute loss value (scalar)
    - `backward(y_true: Tensor, y_pred: Tensor) -> Tensor`: Compute gradient w.r.t. predictions

**Example:**
    ```python
    from neomatrix import losses
    from neomatrix._backend import Tensor
    import numpy as np
    
    loss_fn = losses.HingeLoss()
    
    # Labels MUST be -1 or +1 (NOT 0 or 1!)
    y_true = Tensor.from_numpy(np.array([[-1.0, 1.0, 1.0]]))
    
    # Raw predictions (no activation needed)
    y_pred = Tensor.from_numpy(np.array([[-0.5, 2.0, 0.3]]))
    
    # Forward pass
    loss_val = loss_fn.call(y_true, y_pred)  # Penalizes first and third sample
    
    # Backward pass
    grad = loss_fn.backward(y_true, y_pred)
    ```

**Use Cases:**
    - Binary classification with maximum-margin objective
    - Support Vector Machine (SVM) training
    - When you want predictions to have confidence margin (not just correct sign)

**Properties:**
    - **CRITICAL**: Expects labels in {-1, +1}, NOT {0, 1}
    - No activation function needed on output (raw scores)
    - Zero gradient when prediction is correct with margin ≥ 1
    - Non-zero gradient only for margin violations
    - Encourages confident predictions (not just correct)
    - Range: [0, ∞)

**Limitations:**
    - Only suitable for binary classification
    - Does not produce probability estimates (use BCE + Sigmoid for that)
    - Sensitive to outliers (no saturation like cross-entropy)
    - Requires label conversion: 0 → -1, 1 → +1

**Comparison with BCE:**
    - Hinge: Stops penalizing once margin ≥ 1 (sparse gradients)
    - BCE: Always penalizes, pushing predictions to 0 or 1 (dense gradients)
"""
